### ISSUE TITLE
Export & Import Support

### ISSUE THREAD

**Issue by boramalper** [boramalper/magnetico#187]
Database export & import. =)

**Comment by scriptzteam** [comment]
Aka database export & import? :D

**Comment by boramalper** [comment]
Moving to #197. =)

@ProphetZarquon I don't understand your request exactly. You mean a master-slave kind of database configuration or something else?

**Comment by ProphetZarquon** [comment]
Oh no, identical software, just that different relative resources or geographic locations would cause some instances to benefit greatly by db-sharing from more advantageously positioned or equipped instances. 

The best the more remote or resource-limited instances might be able to offer are rare files (if there's truly a high degree of physical separation between instances), but if a smaller, slower instance happens to pick up one extra torrent from its local network connection, in my book that would be worth it. 

Basically I would hope to see multiple instances of magnetico* running on various hardware at multiple locations, all sharing one db via a DHT. Ideally, a resource-limited instance would not need to retain the entire db in order to search it; rather that cow's query could check the portions of the db held by many cods. Do you believe that might be feasible?

**Comment by boramalper** [comment]
Sounds quite complicated to me, to say that they are sharing one db via a DHT; how different would that be from the current DHT? =)

**Comment by ProphetZarquon** [comment]
I don't imagine that designing & implementing a new DHT is ever easy, unless one already has a good level of experience with DHT protocols... (Their prime use cases seem far more common than developers knowledgeable of them!) 

If you look at the protocol implemented by Delft university's torrent app Tribler, they essentially developed a fork of the existing BitTorrent DHT, to build an index of hashes-having-this-file-metadata (tag(s), filename, size) instead of only an index of peers-having-this-hash. 

The additional features implemented within Tribler's DHT enable Tribler clients to participate in a distributed index of the file-metadata contained within any torrents shared via Tribler clients, making it possible for Tribler clients to search for any filenames (or sizes, tags, etc) indexed by any other Tribler clients. 

Unfortunately, to the best of understanding, this "overlay" DHT is a separate unconnected fork of the 'mainline' public BitTorrent DHT: While Tribler clients still use the BT DHT for locating peers having torrents shared via traditional torrent clients, Tribler clients do not automatically add those torrents to the Tribler DHT; so one can only use the Tribler DHT to search torrents that have been manually added to the Tribler DHT by Tribler clients intending to share that torrent. As a result, most torrents in the BT DHT are not indexed in the Tribler DHT yet (until somebody gets around to setting a Tribler client up with automation to add everything found by a DHT crawler like yours, *cough* *cough*). 

In my (admittedly novice view) the use of a separate Tribler DHT, without crawling the BT DHT to keep the Tribler DHT up to date with all the new torrents in the BT DHT severely limits the utility of Tribler's DHT, & by extension limits the scope of the Tribler client's built-in search functionality. I think that's a mistake that has weakened the appeal of the Tribler client for most users who just want an easy search. Moreover I think it has limited outside developer awareness of the Tribler DHT to such an extent that projects like your own may be duplicating some of the work Delft students & faculty did in implementing a searchable index of publicly available torrents. This means that torrent curation, tagging, accessibility, & lifetime are not as improved by the existence of the Tribler DHT as it could be. 

Nonetheless, the code for the Tribler DHT seems relatively well conceived & the protocol extremely useful, if only more content were being indexed (not many Tribler clients out there, by percentage). Last time I looked their TOR implementation had some flaws but TOR pseudonymity isn't why I'm interested in the Tribler project anyway. Search is the critical defining feature, in my view.

I think a secondary DHT containing hashes-having-this-file-metadata, that runs complementary to the existing peers-having-this-hash BT DHT, with the client _automatically_ adding the metadata of any known torrent, would offer much more utility. 

Similarly, I think the your project could benefit significantly from the reduction in duplicate crawling effort which a metadata DHT could provide, sharing effort among all participating Magnetico^ clients. 

I don't expect that it's trivially easy to implement (especially in an abuse-resistant way, as that's always a challenge with P2P), but I think it could be worth taking a good look at their code to see if the benefits of such an implementation are greater than the difficulty of it: 

https://github.com/Tribler/tribler 
https://www.tribler.org/DistributedTracker
